File: backend/app/schemas/eda.py
"""Pydantic schemas for EDA summaries and AI explanations."""
from typing import Any, Literal

from pydantic import BaseModel

from app.schemas.chart import ChartSpec


class Correlations(BaseModel):
    columns: list[str]
    matrix: list[list[Any]]


class RecommendedChart(BaseModel):
    type: str
    reason: str
    column: str | None = None
    x: str | None = None
    y: str | None = None


class EdaSummary(BaseModel):
    numeric: dict[str, dict[str, Any]]
    correlations: Correlations
    recommended_charts: list[RecommendedChart]


class ExplainRequest(BaseModel):
    kind: Literal["overview", "chart"]
    spec: ChartSpec | None = None


class ExplainResponse(BaseModel):
    text: str
    source: Literal["llm", "fallback"]

File: frontend/src/lib/eda.ts
import { api } from "./api";

export type ChartType =
  | "histogram"
  | "bar"
  | "pie"
  | "box"
  | "scatter"
  | "correlation_heatmap"
  | "line";

export interface ChartSpec {
  type: ChartType;
  column?: string;
  x?: string;
  y?: string;
  bins?: number;
  top_n?: number;
}

export interface ChartData {
  type: ChartType;
  title: string;
  x_label: string | null;
  y_label: string | null;
  categories: unknown[] | null;
  series: { name: string; data: unknown }[];
  extra: Record<string, unknown>;
}

export interface RecommendedChart {
  type: ChartType;
  reason: string;
  column?: string | null;
  x?: string | null;
  y?: string | null;
}

export interface EdaSummary {
  numeric: Record<string, Record<string, number | null>>;
  correlations: { columns: string[]; matrix: (number | null)[][] };
  recommended_charts: RecommendedChart[];
}

export interface ExplainResponse {
  text: string;
  source: "llm" | "fallback";
}

export async function getEdaSummary(id: string): Promise<EdaSummary> {
  const { data } = await api.get<EdaSummary>(`/datasets/${id}/eda/summary`);
  return data;
}

export async function postChart(id: string, spec: ChartSpec): Promise<ChartData> {
  const { data } = await api.post<ChartData>(`/datasets/${id}/chart`, spec);
  return data;
}

export async function explain(
  id: string,
  kind: "overview" | "chart",
  spec?: ChartSpec,
): Promise<ExplainResponse> {
  const { data } = await api.post<ExplainResponse>(`/datasets/${id}/explain`, {
    kind,
    spec: spec ?? null,
  });
  return data;
}
